crud/schema.go

```diff
@@ -0,0 +1,360 @@
+		}
+		*schema = make(map[string]SpaceSchema, l)
+
+		for i := 0; i < l; i++ {
+			key, err := d.DecodeString()
+			if err != nil {
+				return err
+			}
+
+			var spaceSchema SpaceSchema
+			if err := d.Decode(&spaceSchema); err != nil {
+				return err
+			}
+
+			(*schema)[key] = spaceSchema
+		}
+	} else {
+		return fmt.Errorf("unexpected code=%d decoding map or empty array", code)
+	}
+
+	return nil
+}
+
+// SpaceSchema contains a single CRUD space schema definition.
+type SpaceSchema struct {
+	Format  []FieldFormat
+	Indexes map[uint32]Index
+}
+
+// DecodeMsgpack provides custom msgpack decoder.
+func (spaceSchema *SpaceSchema) DecodeMsgpack(d *msgpack.Decoder) error {
+	l, err := d.DecodeMapLen()
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < l; i++ {
+		key, err := d.DecodeString()
+		if err != nil {
+			return err
+		}
+
+		switch key {
+		case "format":
+			la, err := d.DecodeArrayLen()
+			if err != nil {
+				return err
+			}
+
+			spaceSchema.Format = make([]FieldFormat, la)
+			for j := 0; j < la; j++ {
+				if err := d.Decode(&spaceSchema.Format[j]); err != nil {
+					return err
+				}
+			}
+		case "indexes":
+			lm, err := d.DecodeMapLen()
+			if err != nil {
+				return err
+			}
+
+			spaceSchema.Indexes = make(map[uint32]Index, lm)
+			for j := 0; j < lm; j++ {
+				indexId, err := d.DecodeUint32()
+				if err != nil {
+					return err
+				}
+
+				var indexObj Index
+				if err := d.Decode(&indexObj); err != nil {
+					return err
+				}
+
+				spaceSchema.Indexes[indexId] = indexObj
+			}
+		default:
+			if err := d.Skip(); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Index contains a CRUD space index definition.
+type Index struct {
+	Id     uint32
+	Name   string
+	Type   string
+	Unique bool
+	Parts  []IndexPart
+}
+
+// DecodeMsgpack provides custom msgpack decoder.
+func (index *Index) DecodeMsgpack(d *msgpack.Decoder) error {
+	l, err := d.DecodeMapLen()
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < l; i++ {
+		key, err := d.DecodeString()
+		if err != nil {
+			return err
+		}
+
+		switch key {
+		case "id":
+			if index.Id, err = d.DecodeUint32(); err != nil {
+				return err
+			}
+		case "name":
+			if index.Name, err = d.DecodeString(); err != nil {
+				return err
+			}
+		case "type":
+			if index.Type, err = d.DecodeString(); err != nil {
+				return err
+			}
+		case "unique":
+			if index.Unique, err = d.DecodeBool(); err != nil {
+				return err
+			}
+		case "parts":
+			la, err := d.DecodeArrayLen()
+			if err != nil {
+				return err
+			}
+
+			index.Parts = make([]IndexPart, la)
+			for j := 0; j < la; j++ {
+				if err := d.Decode(&index.Parts[j]); err != nil {
+					return err
+				}
+			}
+		default:
+			if err := d.Skip(); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// IndexField contains a CRUD space index part definition.
+type IndexPart struct {
+	Fieldno     uint32
+	Type        string
+	ExcludeNull bool
+	IsNullable  bool
+}
```

Comment:
It seems that the custom `DecodeMsgpack` could be replaced with `msgpack` annotations. It could work without custom the decoding function:
```suggestion
// IndexField contains a CRUD space index part definition.
type IndexPart struct {
	Fieldno     uint32 `msgpack:"fieldno"`
	Type        string  `msgpack:"type"`
	ExcludeNull bool `msgpack:"exclude_null"`
	IsNullable  bool `msgpack:"is_nullable"`
}
```